src/core/server/http/lifecycle/on_post_auth.ts

```diff
@@ -28,90 +27,104 @@ enum ResultType {
   rejected = 'rejected',
 }
 
+interface NextParams {
+  type: ResultType.next;
+}
+
+interface RedirectedParams {
+  type: ResultType.redirected;
+  url: string;
+}
+
+interface RejectedParams {
+  type: ResultType.rejected;
+  error: Error;
+  statusCode?: number;
+}
+
 /** @internal */
-class OnRequestResult {
+class OnPostAuthResult {
```

Comment:
I'm not sure if it's out of scope for this PR or not, but the Spaces `onPostAuth` interceptor also needs to be able to [render a hidden app](https://github.com/elastic/kibana/blob/80aae6bba81f29a1a2533087ecca2003666fc6d0/x-pack/plugins/spaces/server/lib/request_inteceptors/on_post_auth_interceptor.ts#L49-L53) via request takeover.

I'm not sure if the concept of "hidden applications" has been thought through for the new platform or not (@joshdover?), but we'll need equivalent functionality for both Security and Spaces

Reply:
> I'm not sure if the concept of "hidden applications" has been thought through for the new 

no, although I think it's not a blocker. All Spaces related logic can be moved to NP, but LP can call required functionality.
if this case it can be:
```js
const spaces = await newPlatform.plugins.spaces.client.asScoped(request).getAll();
if(spaces.length === 1) ...
if(spaces.length > 0) ...
```

Reply:
I think I agree with that, but we might need additional work to support [dashboard-only-mode's interceptor](https://github.com/elastic/kibana/blob/27e5406d7a493c53a862ba16ae064a61c13819cf/x-pack/plugins/dashboard_mode/server/dashboard_mode_request_interceptor.js#L25)) in the LP.

It's relying on "auth scopes" to determine when to render its hidden app, but auth scopes are identifiers attached to the `request.auth.credentials` object. Will NP's http service allow direct access to such properties in the LP once security moves to NP?

Reply:
@elastic/kibana-security Do we have a desire to keep the “auth scopes” used for dashboard-only-mode? I’m working with Mikhail to make sure the new platform has the necessary functionality to replace `onRequest` and `onPostAuth`. We want to move the authentication logic to the NP ASAP to unblock other plugins from migrating, but part of the auth logic is the concept of scopes, which are used by dashboard only mode in the legacy platform. We don’t have a solution for rendering hidden apps yet (for dashboard mode), and we don’t want to block migrating authc on that if we don’t have to.

Mikhail suggested adding a service in the NP that would allow plugins to associate arbitrary data with a particular request, which would also extend to the LP. I'm thinking we could use that to store the list of roles post-auth in NP, which dashboard-only-mode could use in the LP to determine if it should render its hidden app or not.

It's effectively the same logic, just without the concept of auth scopes. Dasbhoard-only-mode is the only consumer of the auth scopes service, and AFAIK, we don't have plans to use it elsewhere in the near future. It would be one less service to migrate to the NP, too. One of the goals in the NP is to restrict access to auth data server-side. Currently, any plugin can inspect the request headers and auth details, but Mikhail is working on abstracting that out so that plugins don't require direct access anymore.

Reply:
Removing "auth scopes" seems entirely reasonable to me, as long as we can emulate the existing behavior of dashboard-only-mode using the mechanism that you've described. I'm not a huge fan of the way that Hapi does it's auth rules, so I'd much prefer we use our own abstractions.

Reply:
We can extend  core API: 
- registerAuth already receive `credentials` which we can store internally
```js
 const authenticate: AuthenticationHandler = async (req, sessionStorage, t) => {
          if (req.headers.authorization) {
            sessionStorage.set({ value: user, expires: Date.now() + sessionDurationMs });
            return t.authenticated({ credentials: user });
          } else {
            return t.rejected(Boom.unauthorized());
          }
        };
```
- add a method that returns passed credentials and auth status
```typescript
interface Auth {
  status: 'authorized' | 'non-authorized' | 'disabled';
}
HttpCoreSetup {
  ...
  getAuth: (request: KibanaRequest | Request, basePath: string) => Auth
}
```

@spalger why this logic was added in the core ? https://github.com/legrego/kibana/blob/1e6270b61362ebc38215099f2bd811c7be2d936e/src/legacy/ui/ui_render/ui_render_mixin.js#L122

Reply:
Looks good to me!